# <font color = purple>Problem Statement 
<font color = purple>Here we must to make a machine learning model where we can predict the rating of a customer by using there review so that we can validated the rating given by the customer. This will help us indentify case of mismatch and we can remove those mismatched reviews from zomato which will that build customer trust on the platform and helps the user get an accurate picture of the restaurant. 


In [25]:
import pandas as pd
import numpy as np
import re
import nltk

In [2]:
df = pd.read_csv(r'/content/Zomato_reviews.csv')
df.head(2)

rating                                        review_text
0     1.0  Their service is worst, pricing in menu is dif...
1     5.0  really appreciate their quality and timing . I...

In [3]:
df.shape

(27762, 2)

In [4]:
# checking for null values
df.isnull().sum()

rating          0
review_text    14
dtype: int64

#### Dropping null records

In [5]:
df.dropna(inplace = True)

In [6]:
df.isnull().sum()

rating         0
review_text    0
dtype: int64

### <font color = brown > Data Cleaning and Preprocessing

In [7]:
df['review_text']

0        Their service is worst, pricing in menu is dif...
1        really appreciate their quality and timing . I...
2        Went there on a Friday night, the place was su...
3        A very decent place serving good food.\r\nOrde...
4        One of the BEST places for steaks in the city....
                               ...                        
27757    Food quality 4.5/5\r\nHospitality 4/5\r\nManag...
27758    Taste of the food is good and the ambience as ...
27759    Pizza is really thin crust and made from fresh...
27760    Visited last Saturday with my kids ,\r\nIt was...
27761    I love onesta always and their unlimited offer...
Name: review_text, Length: 27748, dtype: object

In [8]:
df['review_text'] = df['review_text'].apply(lambda x: str.lower(x))

In [9]:
df['review_text']

0        their service is worst, pricing in menu is dif...
1        really appreciate their quality and timing . i...
2        went there on a friday night, the place was su...
3        a very decent place serving good food.\r\norde...
4        one of the best places for steaks in the city....
                               ...                        
27757    food quality 4.5/5\r\nhospitality 4/5\r\nmanag...
27758    taste of the food is good and the ambience as ...
27759    pizza is really thin crust and made from fresh...
27760    visited last saturday with my kids ,\r\nit was...
27761    i love onesta always and their unlimited offer...
Name: review_text, Length: 27748, dtype: object

#### Removing \r\n which represents (next line)

In [10]:
df['review_text'] = df['review_text'].apply(lambda x: x.replace('\r\n', ' '))

#### Removing Stopwords

In [11]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [12]:
from nltk.corpus import stopwords
words = stopwords.words('english')

In [13]:
len(words)

179

In [14]:
want = ['no', 'not', 'don', 'won']

In [15]:
for i in want:
    words.remove(i)

In [16]:
len(words)

175

In [17]:
df['review_text'] = df['review_text'].apply(lambda x: ' '.join([i for i in x.split() if i not in words]))

In [18]:
df['review_text']

0        service worst, pricing menu different bill. gi...
1        really appreciate quality timing . tried thatt...
2        went friday night, place surprisingly empty. i...
3        decent place serving good food. ordered chilli...
4        one best places steaks city. tried beef steak ...
                               ...                        
27757    food quality 4.5/5 hospitality 4/5 management ...
27758    taste food good ambience well. need reduce pri...
27759    pizza really thin crust made freshly prepared ...
27760    visited last saturday kids , superb crowd .ver...
27761    love onesta always unlimited offers. really go...
Name: review_text, Length: 27748, dtype: object

In [1]:
### now our stop words are removed now we shall remove puntuations.

#### Remove Punctuation

In [19]:
df['review_text'] = df['review_text'].apply(lambda x: re.sub(r'[^\w\s]', '', x))

In [20]:
df['review_text']

0        service worst pricing menu different bill give...
1        really appreciate quality timing  tried thatti...
2        went friday night place surprisingly empty int...
3        decent place serving good food ordered chilli ...
4        one best places steaks city tried beef steak c...
                               ...                        
27757    food quality 455 hospitality 45 management res...
27758    taste food good ambience well need reduce pric...
27759    pizza really thin crust made freshly prepared ...
27760    visited last saturday kids  superb crowd very ...
27761    love onesta always unlimited offers really goo...
Name: review_text, Length: 27748, dtype: object

#### Lemmatization
-  Applying lemmatization so that the incorectly spelled words are corrected.

In [28]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


True

In [29]:
from nltk.stem import WordNetLemmatizer
  
lemmatizer = WordNetLemmatizer()

In [31]:
df['review_text'] = df['review_text'].apply(lambda x: ' '.join([lemmatizer.lemmatize(i) for i in x.split()]))

#### Removing numbers from reviews
- Removing numbers present in the reviews which is not necssary

In [33]:
df['review_text'] = df['review_text'].apply(lambda x: re.sub(r'[0-9]', '', x))

In [36]:
df['review_text'] = df['review_text'].apply(lambda x: x.replace('  ', ' '))

In [37]:
df['review_text']

0        service worst pricing menu different bill give...
1        really appreciate quality timing tried thattil...
2        went friday night place surprisingly empty int...
3        decent place serving good food ordered chilli ...
4        one best place steak city tried beef steak chi...
                               ...                        
27757    food quality hospitality management response c...
27758    taste food good ambience well need reduce pric...
27759    pizza really thin crust made freshly prepared ...
27760    visited last saturday kid superb crowd very go...
27761    love onesta always unlimited offer really good...
Name: review_text, Length: 27748, dtype: object

In [38]:
df.head()

rating                                        review_text
0     1.0  service worst pricing menu different bill give...
1     5.0  really appreciate quality timing tried thattil...
2     4.0  went friday night place surprisingly empty int...
3     4.0  decent place serving good food ordered chilli ...
4     5.0  one best place steak city tried beef steak chi...

In [39]:
## So our data is cleaned and we can move to next step wich is model bulding.

#### Train Test Split

In [40]:
from sklearn.model_selection import train_test_split

In [41]:
X_train, X_test, Y_train, Y_test = train_test_split(df['review_text'], df['rating'], test_size = 0.3, random_state = 2)
print('X_train', X_train.shape)
print('X_test', X_test.shape)
print('Y_train', Y_train.shape)
print('Y_test', Y_test.shape)

X_train (19423,)
X_test (8325,)
Y_train (19423,)
Y_test (8325,)


In [42]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [43]:
tfidf = TfidfVectorizer(stop_words = 'english', max_features = 5000)
X_train_vec = tfidf.fit_transform(X_train)
X_test_vec = tfidf.transform(X_test)

### <font color = brown > Random Forest (ensembled model)

In [44]:
from sklearn.ensemble import RandomForestRegressor
rfr = RandomForestRegressor(random_state = 5)

In [45]:
rfr.fit(X_train_vec, Y_train)

RandomForestRegressor(random_state=5)

In [46]:
pred_y = rfr.predict(X_test_vec)

In [47]:
pred_y.shape

(8325,)

#### Model Evaluation

In [67]:
from sklearn.metrics import mean_squared_error, r2_score, make_scorer

In [53]:
mse = mean_squared_error(Y_test, pred_y)
print('mse : ', mse)
print('rmse : ', mean_squared_error(Y_test, pred_y, squared = False))
print('r2score : ', r2_score(Y_test, pred_y))

mse :  0.2920834236297081
rmse :  0.5404474291082418
r2score :  0.8261391782186319


#### Grid Search CV

In [59]:
from sklearn.model_selection import GridSearchCV

In [68]:
def MSE(y_true,y_pred):
    mse = mean_squared_error(y_true, y_pred)
    print('MSE: %2.3f' % mse)
    return mse

def R2(y_true,y_pred):    
     r2 = r2_score(y_true, y_pred)
     print('R2: %2.3f' % r2)
     return r2

def two_score(y_true,y_pred):    
    MSE(y_true,y_pred) #set score here and not below if using MSE in GridCV
    score = R2(y_true,y_pred)
    return score

def two_scorer():
    return make_scorer(two_score, greater_is_better=True)

In [69]:
params = {'max_features': ['auto', 'sqrt', 'log2'],
          'max_depth' : [10, 15, 20, 25]}

In [70]:
gcv = GridSearchCV(rfr, cv = 5, param_grid = params, scoring=two_scorer())

In [71]:
gcv.fit(X_train_vec, Y_train)

MSE: 0.973
R2: 0.405
MSE: 0.907
R2: 0.426
MSE: 0.921
R2: 0.446
MSE: 0.891
R2: 0.456
MSE: 0.903
R2: 0.458
MSE: 1.253
R2: 0.234
MSE: 1.207
R2: 0.236
MSE: 1.241
R2: 0.254
MSE: 1.249
R2: 0.238
MSE: 1.255
R2: 0.247
MSE: 1.474
R2: 0.098
MSE: 1.426
R2: 0.098
MSE: 1.499
R2: 0.099
MSE: 1.478
R2: 0.098
MSE: 1.502
R2: 0.099
MSE: 0.777
R2: 0.525
MSE: 0.742
R2: 0.531
MSE: 0.730
R2: 0.561
MSE: 0.726
R2: 0.557
MSE: 0.738
R2: 0.557
MSE: 1.100
R2: 0.327
MSE: 1.070
R2: 0.323
MSE: 1.087
R2: 0.347
MSE: 1.097
R2: 0.331
MSE: 1.105
R2: 0.337
MSE: 1.396
R2: 0.146
MSE: 1.355
R2: 0.143
MSE: 1.421
R2: 0.145
MSE: 1.405
R2: 0.143
MSE: 1.429
R2: 0.143
MSE: 0.654
R2: 0.600
MSE: 0.632
R2: 0.600
MSE: 0.617
R2: 0.629
MSE: 0.623
R2: 0.620
MSE: 0.638
R2: 0.617
MSE: 0.970
R2: 0.407
MSE: 0.948
R2: 0.400
MSE: 0.955
R2: 0.426
MSE: 0.972
R2: 0.407
MSE: 0.974
R2: 0.416
MSE: 1.325
R2: 0.190
MSE: 1.292
R2: 0.183
MSE: 1.353
R2: 0.186
MSE: 1.332
R2: 0.187
MSE: 1.356
R2: 0.186
MSE: 0.575
R2: 0.649
MSE: 0.560
R2: 0.646
MSE: 0.541
R2

GridSearchCV(cv=5, estimator=RandomForestRegressor(random_state=5),
             param_grid={'max_depth': [10, 15, 20, 25],
                         'max_features': ['auto', 'sqrt', 'log2']},
             scoring=make_scorer(two_score))

In [72]:
gcv.best_estimator_

RandomForestRegressor(max_depth=25, random_state=5)

In [73]:
gcv.best_params_

{'max_depth': 25, 'max_features': 'auto'}

In [74]:
gcv.best_score_

0.6591920911457401

#### Creating model with best params

In [75]:
rfr1 = RandomForestRegressor(random_state = 5, max_depth = 25, max_features = 'auto')
rfr1.fit(X_train_vec, Y_train)

RandomForestRegressor(max_depth=25, random_state=5)

In [76]:
pred_y1 = rfr1.predict(X_test_vec)

In [78]:
mse = mean_squared_error(Y_test, pred_y1)
print('mse : ', mse)
print('rmse : ', mean_squared_error(Y_test, pred_y1, squared = False))
print('R2Score : ',r2_score(Y_test, pred_y1))

mse :  0.5534258090580201
rmse :  0.7439259432618411
R2Score :  0.6705767661781841


<font color = olive > ***The score in the cell below if of rfr model which is more accurate score than the score which we achieved by GridSearch CV so we shall choose this model.***

In [53]:
mse = mean_squared_error(Y_test, pred_y)
print('mse : ', mse)
print('rmse : ', mean_squared_error(Y_test, pred_y, squared = False))
print('r2score : ', r2_score(Y_test, pred_y))

mse :  0.2920834236297081
rmse :  0.5404474291082418
r2score :  0.8261391782186319


<font color = olive > ***NOTE: We have just applied GtridSearchCV with the parameter given in the question due to time constrain, but we can increase the R2 score and reduce mse by giving different parameters in GRIdSearch CV. But for now just take this mode with R2 of 82% and mse 0f 0.29.***

# <font color = purple> **Conclusion :** 
  <font color = purple>   So we have created a RandomForestRegressor model predict the rating of a customer by using there review so that we can validated the rating given by the customer. Which will help us to validated the rating given by the customer. This will help us indentify case of mismatch and we can remove those mismatched reviews from zomato which will that build customer trust on the platform and helps the user get an accurate picture of the restaurant.
    